parlai/agents/examples/transformer_variant.py

```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from dataclasses import replace
+import torch
+from typing import Tuple, Optional, Union
+
+from parlai.agents.transformer.modules import (
+    TransformerGeneratorModel,
+    TransformerEncoder,
+)
+from parlai.agents.transformer.transformer import TransformerGeneratorAgent
+
+
+class TransformerVariantAgent(TransformerGeneratorAgent):
+    def build_model(self, states=None):
+        manifest = TransformerGeneratorModel.Manifest()
+        manifest.encoder = replace(manifest.encoder, klass=MyCustomEncoder)
+        return TransformerGeneratorModel(self.opt, self.dict, manifest)
+
+
+class MyCustomEncoder(TransformerEncoder):
+    def forward(  # type: ignore
+        self,
+        input: torch.LongTensor,
+        positions: Optional[torch.LongTensor] = None,
+        segments: Optional[torch.LongTensor] = None,
+    ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.BoolTensor]]:
+        # Comment out the following line and write your custom `forward` instead.
+        return super().forward(input, positions, segments)
```

Comment:
is it convenient to provide an example where the leaf level manifest is swapped: e.g. having a custom MultiHeadAttention in `TransformerDecoderLayer` while the rest (other parts of `TransformerDecoderLayer`, `TransformerGenratorModel`, `Encoder` ...) stay the same.

`replace(TransformerDecoder.Manifest().layer, MyCustomTransformerDecoderLayer)`
where MyCustomTransformerDecoderLayer has manifest.self_attention default value = CustomMultiHeadAttention
or maybe sth like 
`replace(TransformerDecoder.Manifest().layer.Manifest().self_attention,  ???)`

Reply:
Good point, I'll add an example!
You could do either of the options you mentioned, but I had envisioned something more like 2 for simply swapping out the leaf component.